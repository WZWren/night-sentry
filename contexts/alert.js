/**
 * alert.js manages all the PubSub stuff for alerts and services.
 * Functions in this js file will be called back in auth during the login step.
 */

import { supabase } from "../lib/supabase";

export async function fetchPublishers(user, setDistress) {
    const { data, error } = await supabase.from("close_contacts")
        .select("publisher, name:publisher(first_name, last_name)")
        .eq("subscriber", user.id)
        .is("confirmed", true);

    if (error) {
        console.log(error.message);
        return;
    }

    // linear runtime is OK for this, as a user's close contacts should
    // on average be within the 10s range.
    for (let i = 0; i < data.length; i++) {
        // subscribe the app client to the channel.
        supabase.channel(data[i].publisher)
            .on('postgres_changes',
                { event: 'INSERT', schema: 'public', table: 'alerts', filter: `user_id=eq.${data[i].publisher}` },
                (payload) => {
                    setDistress(`Distress signal received from ${data[i].name.first_name} ${data[i].name.last_name}`);
                    console.log(payload);
                }
            ).subscribe();
        console.log("Subscribed to " + data[i].publisher)
    }
}

export async function unsubClient() {
    const { error } = await supabase.removeAllChannels();

    if (error) {
        console.log(error.message);
        return;
    }
    console.log("unsubscribed to all");
}