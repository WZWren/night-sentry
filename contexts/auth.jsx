import { useRouter, useSegments } from "expo-router";

import { createContext, useContext, useEffect, useState } from "react";
import { supabase } from "../lib/supabase";

const AuthContext = createContext({});

function useProtectedRoute(loggedIn) {
    const segments = useSegments();
    const router = useRouter();

    useEffect(() => {
        console.log("Entering protected route...");
        // this checks if the route is in the (auth) group of routes - in case
        // the user is somehow not on this route if the user is not logged in
        const inAuthGroup = segments[0] === "(auth)";
        if (!loggedIn && !inAuthGroup) {
            console.log("Routing to Login...");
            router.replace("/login");
        } else if (loggedIn && inAuthGroup) {
            console.log("Routing to Alert...");
            router.replace("/alert");
        }
    }, [loggedIn, segments, router]);
}

/**
 * Parent component that provides the user value as part of its context.
 * 
 * @param {*} children - The Child Components of the AuthProvider
 * @returns The AuthContext.Provider component, encapsulating the user value.
 */
export function AuthProvider({ children }) {
    const [loggedIn, setLoggedIn] = useState(null);
    console.log("AuthProvider loading...");

    useProtectedRoute(loggedIn);

    // this sets up the useEffect hooks to redirect users to the correct page.
    useEffect(() => {
        const { data } = supabase.auth.onAuthStateChange((event, session) => {
            console.log("User changed! " + event);
            if (event === "SIGNED_IN") {
                setLoggedIn(session.user);
            } else if (event === "SIGNED_OUT") {
                setLoggedIn(null);
            }
        })
        return () => data.subscription.unsubscribe();
    }, []);

    return (
        <AuthContext.Provider value={{ loggedIn }}>{ children }</AuthContext.Provider>
    );
}