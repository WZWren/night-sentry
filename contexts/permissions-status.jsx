export const LocalPermStatus = Object.freeze({
    INIT: 'init',
    GRANTED: 'granted',
    UNDETERMINED: 'undetermined',
    DENIED: 'denied',
});