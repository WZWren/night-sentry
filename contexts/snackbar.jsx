import { createContext, useContext, useState } from "react";
import { Snackbar } from "react-native-paper";

const SnackbarContext = createContext({});

export function useSnackbar() {
    return useContext(SnackbarContext);
}

/**
 * An abstraction of the Snackbar element used to display error/alert messages.
 * 
 * @param {*} children The Child components of SnackbarProvider
 * @returns The SnackbarContext.Provider encapsulating the setMessage hook.
 */
export function SnackbarProvider({ children }) {
    const [message, setMessage] = useState("");
    console.log("SnackbarProvider loading...");

    const handleDismiss = () => {
        setMessage("");
    }

    return (
        <SnackbarContext.Provider value={{ setMessage }}>
            { children }
            <Snackbar
                visible={message}
                onDismiss={handleDismiss}
                duration={10000}
                action={{
                    label: "Close",
                    onPress: handleDismiss
                }}>
                {message}
            </Snackbar>
        </SnackbarContext.Provider>
    );
}