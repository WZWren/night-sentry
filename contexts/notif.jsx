// TODO: Move Notifications to its separate context.

import { createContext, useContext, useEffect, useState, useRef } from "react";
import { Platform } from "react-native";
import * as Device from 'expo-device';
import * as Notifications from 'expo-notifications';

import { supabase } from "../lib/supabase";
import { useAuth } from "./auth";

const NotificationsContext = createContext({});

export function useNotif() {
    return useContext(NotificationsContext);
}

// we handle the notification signup in the layout slot.
// notifications taken directly from the expo-notifications example.
Notifications.setNotificationHandler({
    handleNotification: async () => ({
        shouldShowAlert: true,
        shouldPlaySound: true,
        shouldSetBadge: false,
    }),
});

async function registerForPushNotificationsAsync() {
    let token;

    if (Platform.OS === 'android') {
        await Notifications.setNotificationChannelAsync('default', {
            name: 'default',
            importance: Notifications.AndroidImportance.MAX,
            vibrationPattern: [0, 250, 250, 250],
            lightColor: '#FF231F7C',
        });
    }

    if (Device.isDevice) {
        const { status: existingStatus } = await Notifications.getPermissionsAsync();
        let finalStatus = existingStatus;
        if (existingStatus !== 'granted') {
            const { status } = await Notifications.requestPermissionsAsync();
            finalStatus = status;
        }
        if (finalStatus !== 'granted') {
            console.log('Failed to get push token for push notification!');
            return;
        }
        token = (await Notifications.getExpoPushTokenAsync()).data;
        console.log(token);
    } else {
        console.log('Must use physical device for Push Notifications');
    }

    return token;
}

export function NotificationsProvider({ children }) {
    const { loggedIn } = useAuth();
    // expoPushToken is to be pushed to the supabase user_info table.
    const [ expoPushToken, setExpoPushToken ] = useState("");
    // notification is the physical notification from the app
    const [ notification, setNotification ] = useState(null);
    const notificationListener = useRef();
    const responseListener = useRef();

    useEffect(() => {
        registerForPushNotificationsAsync().then(token => setExpoPushToken(token));
    
        notificationListener.current = Notifications.addNotificationReceivedListener(notification => {
            setNotification(notification);
            console.log("Notification received! " + notification.request.content.title);
        });
    
        responseListener.current = Notifications.addNotificationResponseReceivedListener(response => {
            // TODO: this is where router linking will take place.
            console.log(response);
        });
    
        return () => {
            Notifications.removeNotificationSubscription(notificationListener.current);
            Notifications.removeNotificationSubscription(responseListener.current);
        };
    }, []);

    useEffect(() => {
        if (loggedIn && expoPushToken) {
            console.log("Updating user token: " + expoPushToken);
            (async () => {
                const { error } = await supabase.from("user_info")
                    .update({ expo_token: expoPushToken })
                    .eq("id", loggedIn.id);
                if (error) {
                    console.log(error.message);
                }
            })();
        }
    }, [loggedIn, expoPushToken]);

    return (
        <NotificationsContext.Provider value={{ notification, setNotification, expoPushToken }}>
            {children}
        </NotificationsContext.Provider>
    );
}