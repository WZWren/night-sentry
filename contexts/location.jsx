import { useState, useEffect, createContext, useContext } from 'react';
import * as Location from 'expo-location';

import { LocalPermStatus } from './permissions-status';

const LocationContext = createContext({});

export function useLocation() {
    return useContext(LocationContext);
}

export function LocationProvider({children}) {
    const [permissionStatus, setPermissionStatus] = useState(null);
    const [location, setLocation] = useState(null);

    useEffect(() => {
        if (permissionStatus == LocalPermStatus.INIT) {
            (async () => {
                let { status } = await Location.requestForegroundPermissionsAsync();
                setPermissionStatus(status);
                if (status !== LocalPermStatus.GRANTED) {
                  console.log('Permission to access location was denied/not accepted yet.');
                  return;
                }
                let location = await Location.getCurrentPositionAsync({});
                setLocation(location);
                console.log(location.coords.latitude + ", " + location.coords.longitude);
            })();
        }
    }, [permissionStatus]);

    return (
        <LocationContext.Provider value={{ location, permissionStatus, setPermissionStatus }}>
            {children}
        </LocationContext.Provider>
    );
}