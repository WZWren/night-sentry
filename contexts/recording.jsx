import { createContext, useContext, useEffect, useState } from "react";
import { Audio } from "expo-av";
import { decode } from 'base64-arraybuffer';
import * as Notifications from 'expo-notifications';
import * as FileSystem from "expo-file-system";

import { useAuth } from "./auth";
import { supabase } from "../lib/supabase";
import { LocalPermStatus } from "./permissions-status";

const RecorderContext = createContext({});

export function useRecorder() {
    return useContext(RecorderContext);
}

export function RecorderProvider({children}) {
    const { loggedIn } = useAuth();
    const [ recording, setRecording ] = useState(null);
    const [ permissionStatus, setPermissionStatus ] = useState(null);

    useEffect(() => {
        if (permissionStatus == LocalPermStatus.INIT) {
            (async () => {
                let { status } = await Audio.requestPermissionsAsync();
                setPermissionStatus(status);
                if (status !== LocalPermStatus.GRANTED) {
                  console.log('Permission to access microphone was denied/not accepted yet.');
                  return;
                }
            })();
        }
    }, [permissionStatus])

    async function startRecording() {
        try {
            console.log('Requesting permissions..');
            await Audio.requestPermissionsAsync();
            await Audio.setAudioModeAsync({
                allowsRecordingIOS: true,
                playsInSilentModeIOS: true,
            });
    
            console.log('Starting recording..');
            const { recording } = await Audio.Recording.createAsync(Audio.RecordingOptionsPresets.HIGH_QUALITY);
            await Notifications.scheduleNotificationAsync({
                content: {
                    title: "Recording in progress...",
                    subtitle: "Tap to go to alerts page.",
                    autoDismiss: true,
                },
                identifier: "recording",
                trigger: null
            });
            setRecording(recording);
            console.log('Recording started');
        } catch (err) {
            console.error('Failed to start recording', err);
        }
    }

    async function stopRecording() {
        console.log('Stopping recording..');
        setRecording(undefined);
        await recording.stopAndUnloadAsync();
        await Audio.setAudioModeAsync({
          allowsRecordingIOS: false,
        });
        Notifications.dismissNotificationAsync("recording");
        const uri = recording.getURI();
        console.log('Recording stopped and stored at', uri);
        const file = await FileSystem.readAsStringAsync(uri, { encoding: FileSystem.EncodingType.Base64 });
        const { data, error } = await supabase.storage.from("audio").upload(`${loggedIn.id}/${Date.now()}.m4a`, decode(file), {
            contentType: 'audio/mpeg'
        });
        if (error) {
            console.log("failed to upload. " + error.message);
        }
        console.log(data);
    }

    return (
        <RecorderContext.Provider value={{permissionStatus, setPermissionStatus, startRecording, stopRecording, recording}}>
            {children}
        </RecorderContext.Provider>
    );
}