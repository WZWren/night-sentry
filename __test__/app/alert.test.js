import { Text, View } from "react-native";
import { screen, render, act, fireEvent } from "@testing-library/react-native";
import { renderRouter } from "expo-router/src/testing-library";

import * as Location from "../../contexts/location";
import * as Auth from "../../contexts/auth";
import * as Recording from "../../contexts/recording";
import AlertPage from "../../app/(main)/alert";

jest.useFakeTimers();

it("should render properly", () => {
    const renderPage = render(<AlertPage />);
    const distressBtn = renderPage.findByText("Distress Signal");
    const recordingLabel = renderPage.findByText("Tap to Record");
    expect(distressBtn).toBeTruthy();
    expect(recordingLabel).toBeTruthy();

    // expect the basic elements on the alert page to be present - if it is, proceed
    // with snapshot
    const tree = renderPage.toJSON();
    expect(tree).toMatchSnapshot();
});