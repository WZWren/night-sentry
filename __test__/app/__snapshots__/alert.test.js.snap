// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render properly 1`] = `
<View
  style={
    {
      "alignItems": "center",
      "flex": 1,
      "gap": 4,
      "justifyContent": "center",
    }
  }
>
  <View
    style={
      {
        "minHeight": 40,
        "width": "80%",
      }
    }
  >
    <Text
      style={
        [
          {
            "textAlign": "left",
          },
          {
            "color": "rgba(28, 27, 31, 1)",
            "writingDirection": "ltr",
          },
          [
            {
              "fontFamily": "System",
              "fontSize": 24,
              "fontWeight": "400",
              "letterSpacing": 0,
              "lineHeight": 32,
            },
            {
              "textAlign": "center",
            },
          ],
        ]
      }
      testID="status_msg"
    >
      Send alert to your registered close contacts.
    </Text>
  </View>
  <View
    collapsable={false}
    style={
      {
        "backgroundColor": "#101010",
        "borderRadius": 20,
        "shadowColor": "#000",
        "shadowOffset": {
          "height": 0,
          "width": 0,
        },
        "shadowOpacity": 0,
        "shadowRadius": 0,
      }
    }
    testID="button-container-outer-layer"
  >
    <View
      collapsable={false}
      style={
        {
          "backgroundColor": "#101010",
          "borderColor": "transparent",
          "borderRadius": 20,
          "borderStyle": "solid",
          "borderWidth": 0,
          "flex": undefined,
          "minWidth": 64,
          "shadowColor": "#000",
          "shadowOffset": {
            "height": 0,
            "width": 0,
          },
          "shadowOpacity": 0,
          "shadowRadius": 0,
        }
      }
      testID="button-container"
    >
      <View
        accessibilityRole="button"
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": false,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          [
            {
              "overflow": "hidden",
            },
            {
              "borderRadius": 20,
            },
          ]
        }
        testID="button"
      >
        <View
          style={
            [
              {
                "alignItems": "center",
                "flexDirection": "row",
                "justifyContent": "center",
              },
              {
                "height": 250,
              },
            ]
          }
        >
          <Text
            numberOfLines={1}
            selectable={false}
            style={
              [
                {
                  "textAlign": "left",
                },
                {
                  "color": "rgba(28, 27, 31, 1)",
                  "writingDirection": "ltr",
                },
                [
                  {
                    "fontFamily": "System",
                    "fontSize": 14,
                    "fontWeight": "500",
                    "letterSpacing": 0.1,
                    "lineHeight": 20,
                  },
                  [
                    {
                      "marginHorizontal": 16,
                      "marginVertical": 9,
                      "textAlign": "center",
                    },
                    false,
                    {
                      "marginHorizontal": 24,
                      "marginVertical": 10,
                    },
                    undefined,
                    false,
                    {
                      "color": "#a0a0a0",
                      "fontFamily": "System",
                      "fontSize": 14,
                      "fontWeight": "500",
                      "letterSpacing": 0.1,
                      "lineHeight": 20,
                    },
                    {
                      "fontSize": 30,
                      "padding": 12,
                    },
                  ],
                ],
              ]
            }
            testID="button-text"
          >
            Distress Signal
          </Text>
        </View>
      </View>
    </View>
  </View>
  <View
    style={
      {
        "alignItems": "center",
        "flexDirection": "row",
        "gap": 4,
        "padding": 8,
      }
    }
  >
    <View
      style={
        {
          "flex": 1,
        }
      }
    />
    <View
      collapsable={false}
      style={
        {
          "backgroundColor": "transparent",
          "borderRadius": 32,
          "flex": 1,
          "height": 64,
          "margin": 6,
          "shadowColor": "#000",
          "shadowOffset": {
            "height": 0,
            "width": 0,
          },
          "shadowOpacity": 0,
          "shadowRadius": 0,
          "width": 64,
        }
      }
      testID="recordBtn-container-outer-layer"
    >
      <View
        collapsable={false}
        style={
          {
            "backgroundColor": "transparent",
            "borderColor": "rgba(121, 116, 126, 1)",
            "borderRadius": 32,
            "borderWidth": 1,
            "elevation": 0,
            "flex": 1,
            "overflow": "hidden",
            "shadowColor": "#000",
            "shadowOffset": {
              "height": 0,
              "width": 0,
            },
            "shadowOpacity": 0,
            "shadowRadius": 0,
          }
        }
        testID="recordBtn-container"
      >
        <View
          accessibilityComponentType="button"
          accessibilityRole="button"
          accessibilityState={
            {
              "busy": undefined,
              "checked": undefined,
              "disabled": false,
              "expanded": undefined,
              "selected": undefined,
            }
          }
          accessibilityTraits="button"
          accessibilityValue={
            {
              "max": undefined,
              "min": undefined,
              "now": undefined,
              "text": undefined,
            }
          }
          accessible={true}
          centered={true}
          collapsable={false}
          focusable={true}
          hitSlop={
            {
              "bottom": 6,
              "left": 6,
              "right": 6,
              "top": 6,
            }
          }
          onBlur={[Function]}
          onClick={[Function]}
          onFocus={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            [
              {
                "overflow": "hidden",
              },
              [
                {
                  "alignItems": "center",
                  "flexGrow": 1,
                  "justifyContent": "center",
                },
                {
                  "borderRadius": 32,
                },
              ],
            ]
          }
          testID="recordBtn"
        >
          <Text
            accessibilityElementsHidden={true}
            allowFontScaling={false}
            importantForAccessibility="no-hide-descendants"
            pointerEvents="none"
            selectable={false}
            style={
              [
                {
                  "color": "#880000",
                  "fontSize": 48,
                },
                [
                  {
                    "lineHeight": 48,
                    "transform": [
                      {
                        "scaleX": 1,
                      },
                    ],
                  },
                  {
                    "backgroundColor": "transparent",
                  },
                ],
                {
                  "fontFamily": "material-community",
                  "fontStyle": "normal",
                  "fontWeight": "normal",
                },
                {},
              ]
            }
          >
            ó°‘Š
          </Text>
        </View>
      </View>
    </View>
    <View
      style={
        {
          "alignItems": "center",
          "flex": 3,
          "gap": 12,
          "justifyContent": "center",
        }
      }
    >
      <View
        accessibilityRole="progressbar"
        accessibilityState={
          {
            "busy": true,
          }
        }
        accessibilityValue={
          {
            "max": 100,
            "min": 0,
            "now": 0,
          }
        }
        accessible={true}
        onLayout={[Function]}
      >
        <View
          collapsable={false}
          style={
            {
              "backgroundColor": "rgba(231, 224, 236, 1)",
              "height": 4,
              "opacity": 0,
              "overflow": "hidden",
              "width": 150,
            }
          }
        />
      </View>
      <Text
        style={
          [
            {
              "textAlign": "left",
            },
            {
              "color": "rgba(28, 27, 31, 1)",
              "writingDirection": "ltr",
            },
            [
              {
                "fontFamily": "System",
                "fontSize": 14,
                "fontWeight": "500",
                "letterSpacing": 0.1,
                "lineHeight": 20,
              },
              undefined,
            ],
          ]
        }
      >
        Tap to Record
      </Text>
    </View>
    <View
      style={
        {
          "flex": 1,
        }
      }
    />
  </View>
  <Text
    style={
      [
        {
          "textAlign": "left",
        },
        {
          "color": "rgba(28, 27, 31, 1)",
          "writingDirection": "ltr",
        },
        [
          {
            "fontFamily": "System",
            "fontSize": 24,
            "fontWeight": "400",
            "letterSpacing": 0,
            "lineHeight": 32,
          },
          undefined,
        ],
      ]
    }
  >
    You can send an alert from a homescreen widget too.
  </Text>
</View>
`;
